### PR TITLE
Skip redundant registration for Fonts API-registered fonts

### PR DESCRIPTION
**Pull request by ironprogrammer**
## What?
This update synchronizes the origin name checked for fonts registered through the Fonts API. Those fonts are already registered, so can be skipped here.

## Why?
Addresses #48263, where all fonts registered through a plugin are rendered on the front end, regardless of whether they are in use by global, template, or per-page style settings.

## How?
Updates the origin name check to `gutenberg_wp_fonts_api`, to [match where it's being set originally](https://github.com/WordPress/gutenberg/blob/37bb7c59f032afd034c136db18456388bb45c1fe/lib/experimental/fonts-api/class-wp-fonts.php#L734).

## Testing Instructions
_If you have previously set custom fonts for global styles, templates/patterns through the editor, or on a per-post basis, please reset the typography to their defaults for the active template, otherwise test results may vary._

1. On your test site, activate the TT3 theme, as well as the Gutenberg plugin (from this branch).
2. Install and activate [this test plugin](https://github.com/ironprogrammer/webfonts-jetpack-test).
3. View the site frontend, and view source or inspect the elements/markup.
4. 👀 Near the end of the `<head>` element, there should be a `<style id="wp-fonts-local">` tag, but there should NOT be a `<style id="wp-fonts-jetpack-google-fonts">` tag.
5. In WP admin, create a new post, and add a heading.
6. Open the Settings sidebar, and under Typography enable Font Family. The added Font dropdown should list numerous fonts registered by the test plugin.
7. Set the heading's Font to one of the non-default font options (e.g. Arvo), and Publish or Update the post.
8. View the post on the site frontend, and view source or inspect the elements/markup.
9. 👀 Near the end of the `<head>` element, there should now be a `<style id="wp-fonts-jetpack-google-fonts">` tag, and the `@font-face` declarations should match the custom font selected above. For instance, for Arvo there should be 4 declarations.
10. In WP admin, navigate to _Appearance > Editor_, and click anywhere on the right side to go into edit mode.
11. Open the Styles panel, and select Typography.
12. Under Elements select Headings, and set the Font to something other than a default TT3 font or the one selected above (e.g. Space Mono), then Save the custom styles.
13. View the previous post on the site frontend, and view source or inspect the elements/markup.
14. 👀 The `<style id="wp-fonts-jetpack-google-fonts">` tag should include `@font-face` declarations for both the custom font selected on the post, as well as the font set in the global styles above. For instance, for Arvo there should be 4 declarations, and Space Mono should have 12.

### Expected Results
- ✅ With the test plugin activated, but no custom fonts assigned, unassigned fonts should not be output on the site frontend (i.e. no `<style id="wp-fonts-jetpack-google-fonts">` tag).
- ✅ With a custom font(s) assigned in a post, the `@font-face`(s) should render on the site frontend in the `<style id="wp-fonts-jetpack-google-fonts">` tag.
- ✅ With a custom font(s) assigned globally, the `@font-face`(s) should render on the site frontend's `<style id="wp-fonts-jetpack-google-fonts">` tag, in addition to other custom fonts per template or page.

Please note that logged deprecation notices for `wp_webfonts` and `wp_enqueue_webfont` are expected.

### Testing Instructions for Keyboard
n/a

## Screenshots or screencast
n/a

## Additional Note
[The test plugin provided for](https://github.com/mreishus/webfonts-jetpack-test) #48263 includes registration of fonts already present in TT3, and results in those fonts being doubly-rendered in the front end (first in `#wp-fonts-local`, and then `#wp-fonts-jetpack-google-fonts`). The [updated test plugin for this PR](https://github.com/ironprogrammer/webfonts-jetpack-test) mitigates that issue by removing those fonts, as that bug should be tracked down separately.

### CODE REVIEW

File: lib/experimental/fonts-api/register-fonts-from-theme-json.php [hunk 1]
```diff
@@ -59,8 +59,8 @@ function gutenberg_register_fonts_from_theme_json() {
 				$font_family['fontFace'] = (array) $font_family['fontFace'];
 
 				foreach ( $font_family['fontFace'] as $font_face ) {
-					// Skip if the webfont was registered through the Webfonts API.
-					if ( isset( $font_face['origin'] ) && 'gutenberg_wp_webfonts_api' === $font_face['origin'] ) {
+					// Skip if the webfont was registered through the Fonts API.
```
Comment (hellofromtonya) [inline lib/experimental/fonts-api/register-fonts-from-theme-json.php]:
```suggestion
					// Skip if the font was registered through the Fonts API.
```

File: lib/experimental/fonts-api/deprecations/class-wp-web-fonts.php [hunk 2]
```diff
@@ -690,7 +690,7 @@ public function to_theme_json( $font_family_handle ) {
 			}
 
 			$variation_obj        = $this->registered[ $variation_handle ];
-			$variation_properties = array( 'origin' => 'gutenberg_wp_webfonts_api' );
+			$variation_properties = array( 'origin' => 'gutenberg_wp_fonts_api' );
```
Comment (hellofromtonya) [inline lib/experimental/fonts-api/deprecations/class-wp-web-fonts.php:693]: Is there risk of this impacting extenders who may be iterating through the registered fonts? 

A search of the wporg plugin directory shows only Gutenberg using it https://wpdirectory.net/search/01GSZ95M6F3772YGWKZXG9A15P.

I suspect the risk to users is minimally low.

**Review by hellofromtonya — APPROVED** [review]
tl;dr
This PR resolves the issue and is good for merge 👍 

Reasoning:

* Before applying the PR, can reproduce the issue ✅ 
* After applying the PR, resolves the issue ✅ 
* The changes to the BC Layer have minimal risk of breakages: the `'origin'` is an internal flag and there are no instances of its usage in wporg plugin directory.

Needs tests to avoid this kind of regression, but thinking these tests can be added separately as more thought is needed for the coverage paths.

File: lib/experimental/fonts-api/deprecations/class-wp-web-fonts.php (hunk 2 above)
Comment (zaguiini) [reply lib/experimental/fonts-api/deprecations/class-wp-web-fonts.php:693]: If anything, this feature is in the `experimental` folder so external developers should be aware that things might break 😅 

Is it worth it to include a `doing_it_wrong` call to announce that things changed, @hellofromtonya?

Comment (hellofromtonya) [reply lib/experimental/fonts-api/deprecations/class-wp-web-fonts.php:693]: >If anything, this feature is in the experimental folder so external developers should be aware that things might break 😅
True, though the API now includes a temporary BC Layer which is intended to keep sites running while giving time for extenders to upgrade to the new API.

>Is it worth it to include a doing_it_wrong call to announce that things changed, @hellofromtonya?

Good question @zaguiini! The origin is set internally within the API itself. So I think it's okay to not add a `doing_it_wrong` here. But it does me think about: could it be set externally? If yes and set wrong, what happens? Hmm that's interesting as it may cause unexpected things. Thinking a new issue is needed to discuss and explore.

File: lib/experimental/fonts-api/register-fonts-from-theme-json.php [hunk 3]
```diff
@@ -59,8 +59,8 @@ function gutenberg_register_fonts_from_theme_json() {
 				$font_family['fontFace'] = (array) $font_family['fontFace'];
 
 				foreach ( $font_family['fontFace'] as $font_face ) {
-					// Skip if the webfont was registered through the Webfonts API.
-					if ( isset( $font_face['origin'] ) && 'gutenberg_wp_webfonts_api' === $font_face['origin'] ) {
+					// Skip if the font was registered through the Fonts API.
+					if ( isset( $font_face['origin'] ) && 'gutenberg_wp_fonts_api' === $font_face['origin'] ) {
```
Comment (hellofromtonya) [inline lib/experimental/fonts-api/register-fonts-from-theme-json.php:63]: Update: Never mind : origin is not saved in the database.

Thinking a little more about this ...

User selected fonts are stored in the database, i.e. in the `'wp_global_styles'` custom post type. This means the origin could already exist in the database. Thus, this won't work for saved global style typography.

Instead, will need to check for either `'gutenberg_wp_fonts_api'` OR `'gutenberg_wp_webfonts_api'`.

```suggestion
					if ( 
						isset( $font_face['origin'] ) && 	
						(
							'gutenberg_wp_fonts_api' === $font_face['origin'] ||
							'gutenberg_wp_webfonts_api' === $font_face['origin']
						)
					) {
```

**Comment by pbking** [comment]
I can confirm the change fixes the problem and doesn't negatively impact Jetpack's or Blockbase's font stuff.

File: lib/experimental/fonts-api/register-fonts-from-theme-json.php (hunk 3 above)
Comment (hellofromtonya) [reply lib/experimental/fonts-api/register-fonts-from-theme-json.php:63]: Never mind. On further testing, `'origin'` is not saved to the database. That's great news as it eliminates the need for a migration plan when this API is introduced into WordPress Core 😅

**Comment by hellofromtonya** [comment]
@DaisyOlsen this patch is needed in 15.2.1 release (see https://github.com/WordPress/gutenberg/issues/48263).
